README.md
# tcp_socket_connection

This is a Flutter package that allows you to connect to a socket over the net. All data is then read using UTF-8.

## Getting started

To use this plugin, add *tcp_socket_connection* as a dependency in your pubspec.yaml file.

## Main Example
Here is shown an example where a connection is established and the received message is displayed in a Text widget.
```javascript
import 'package:flutter/material.dart';
import 'package:tcp_socket_connection/tcp_socket_connection.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
        visualDensity: VisualDensity.adaptivePlatformDensity,
      ),
      home: MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);
  final String title;
  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  //Instantiating the class with the Ip and the PortNumber
  TcpSocketConnection socketConnection=TcpSocketConnection("192.168.1.113", 10251);

  String message="";

  @override
  void initState() {
    super.initState();
    startConnection();
  }

  //receiving and sending back a custom message
  void messageReceived(String msg){
    setState(() {
      message=msg;
    });
    socketConnection.sendMessage("MessageIsReceived :D ");
  }

  //starting the connection and listening to the socket asynchronously
  void startConnection() async{
    socketConnection.enableConsolePrint(true);    //use this to see in the console what's happening
    if(await socketConnection.canConnect(5000, attempts: 3)){   //check if it's possible to connect to the endpoint
      await socketConnection.connect(5000, messageReceived, attempts: 3);
    }
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Center(
        child: Text(
          'You have received '+ message,
        ),
      ),
    );
  }
}
```
## Usage
First of all declare a new instance of the *TcpSocketConnection* class:
```javascript
TcpSocketConnection socketConnection=TcpSocketConnection(IP_ADDRESS, PORT_NUMBER);
```
Then, to establish a connection it's used the following method:

```javascript
await socketConnection.connect(TIMEOUT, CALLBACK_FUNCTION);
```

The method <span style="color:navy">**connect**</span> allows you to set a TIMEOUT before finishing attempting to connect, a CALLBACK_FUNCTION that is created by the developer and takes a <span style="color:darkred">**String**</span> as parameter. The CALLBACK_FUNCTION will be called when the message is received.

The word **await** is really important since the connection requires a little bit of time to be initialized. Write that line of code inside an async method to ensure it doesn't impact the UI.

To send a message use the following method:
```javascript
socketConnection.sendMessage(MESSAGE);
```
In case you want to send a message with also an EOM you can directly use the following method:
```javascript
socketConnection.sendMessageEOM(MESSAGE,EOM);
```
It will automatically append the EOM (End Of Message) at the end of the MESSAGE.

## Examples
Imagine you are receiving a <span style="color:darkred">**String**</span> from a server which states *"Confirm?FINISH"*. In order to read it you have to listen to the server asynchronously until you receive the EOM (End of Message) to be sure that the message is finished. After that you have to split the <span style="color:darkred">**String**</span> and keep only the part you need, in this case *"Confirm?"*. This is all done by the <span style="color:navy">**connectEOM**</span> method in only 1 line of code!
```javascript
await socketConnection.connectEOM(TIMEOUT, EOM, CALLBACK_FUNCTION, {ATTEMPTS=1});
```
Imagine you are receiving a <span style="color:darkred">**String**</span> from a server.  With the <span style="color:navy">**connect**</span> method you can read messages received (that probably need to be elaborated by you: the developer) in only 1 line of code!
```javascript
await socketConnection.connect(TIMEOUT, CALLBACK_FUNCTION, {ATTEMPTS=1});
```

## Last but not least

If you find this package useful, don't forget to leave a like! I appreciate it!


CHANGELOG.md
## [0.3.0] - 24/05/2022 - Removed methods and improved performance
Removed useless methods and drastically improved performance by means of StringBuffers.

## [0.2.2] - 08/04/2021 - Fixed bug
Fixed NullPointerException bug when using simpleConnect

## [0.2.1] - 13/12/2020 - Added attempts and 'canConnect' method
Added optional parameter 'attempts' to specify how many times the socket should retry to connect before stopping.

## [0.2.0] - 13/12/2020 - Added attempts and 'canConnect' method
Added optional parameter 'attempts' to specify how many times the socket should retry to connect before stopping.

## [0.1.0] - 13/12/2020 - Added some functionalities
Added some try/catch in order to handle exception when it's impossible to connect to the endpoint.

## [0.0.10] - 13/12/2020 - Added some functionalities
Now it's possible to use the method 'simpleConnect' to read received messages "as they are". The developer elaborates them as he wants.

## [0.0.8] - 01/12/2020 - Added some functionalities
Now you can choose to print events in the console automatically through the method 'enableConsolePrint'

## [0.0.7] - 26/11/2020 - Modified the README file.
Modified some variable names.




test/tcp_socket_connection_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:tcp_socket_connection/tcp_socket_connection.dart';

void main() {
  /*
  test('adds one to input values', () {
    final calculator = Calculator();
    expect(calculator.addOne(2), 3);
    expect(calculator.addOne(-7), -6);
    expect(calculator.addOne(0), 1);
    expect(() => calculator.addOne(null), throwsNoSuchMethodError);
  });*/
}


lib/tcp_socket_connection.dart
library tcp_socket_connection;
import 'dart:async';
import 'dart:io';
import 'dart:convert';

class TcpSocketConnection{

  String _ipAddress;
  int _portAddress;
  Socket _server;
  bool _connected=false;
  bool _logPrintEnabled=false;

  /// Initializes che class itself
  ///  * @param  ip  the server's ip you are trying to connect to
  ///  * @param  port the servers's port you are trying to connect to
  TcpSocketConnection(String ip, int port){
    _ipAddress=ip;
    _portAddress=port;
  }

  /// Initializes che class itself
  ///  * @param  the ip  server's ip you are trying to connect to
  ///  * @param  the port servers's port you are trying to connect to
  ///  * @param  enable if set to true, then events will be printed in the console
  TcpSocketConnection.constructorWithPrint(String ip, int port, bool enable){
    _ipAddress=ip;
    _portAddress=port;
    _logPrintEnabled=enable;
  }

  /// Shows events in the console with print method
  /// * @param  enable if set to true, then events will be printed in the console
  enableConsolePrint(bool enable){
    _logPrintEnabled=enable;
  }

/*
  /// Initializes the connection. Socket starts listening to server for data.
  /// 'callback' function will be called when 'eom' is received
  ///
  ///  * @param  timeOut  amount of time to attempt the connection in milliseconds
  ///  * @param  separator  sequence of characters to use between commands
  ///  * @param  eom  sequence of characters at the end of each single message
  ///  * @param  callback  function called when received a message. It must take 2 'String' as params. The first one is the command received, the second one is the message itself.
  ///  * @param  attempts  number of attempts before stop trying to connect. Default is 1.
  connectWithCommand(int timeOut, String separator, String eom,  Function callback, {int attempts=1}) async{
    if(_ipAddress==null){
      print("Class not initialized. You must call the constructor!");
      return;
    }
    _eom=eom;
    _separator=separator;
    int k=1;
    while(k<=attempts){
      try{
        _server = await Socket.connect(_ipAddress, _portAddress, timeout: new Duration(milliseconds: timeOut));
        break;
      }catch(Exception){
        _printData(k.toString()+" attempt: Socket not connected (Timeout reached)");
        if(k==attempts){
          return;
        }
      }
      k++;
    }
    _connected=true;
    _printData("Socket successfully connected");
    String message="";
    _server.listen((List<int> event) async {
      message += (utf8.decode(event));
      if(message.contains(eom)){
        List<String> commands=message.split(_separator);
        _printData("Message received: "+message);
        callback(commands[0],commands[1].split(eom)[0]);
        if(commands[1].split(eom).length>1){
          message=commands[1].split(eom)[1];
        }else{
          message="";
        }
      }
    });
  }*/

  /// Initializes the connection. Socket starts listening to server for data
  /// 'callback' function will be called whenever data is received. The developer elaborates the message received however he wants
  /// No separator is used to split message into parts
  ///  * @param  timeOut  the amount of time to attempt the connection in milliseconds
  ///  * @param  callback  the function called when received a message. It must take a 'String' as param which is the message received
  ///  * @param  attempts  the number of attempts before stop trying to connect. Default is 1.
  connect(int timeOut, Function callback,{int attempts=1}) async{
    if(_ipAddress==null){
      print("Class not initialized. You must call the constructor!");
      return;
    }
    int k=1;
    while(k<=attempts){
      try{
        _server = await Socket.connect(_ipAddress, _portAddress, timeout: new Duration(milliseconds: timeOut));
        break;
      }catch(Exception){
        _printData(k.toString()+" attempt: Socket not connected (Timeout reached)");
        if(k==attempts){
          return;
        }
      }
      k++;
    }
    _connected=true;
    _printData("Socket successfully connected");
    _server.listen((List<int> event) async {
      String received=(utf8.decode(event));
      _printData("Message received: "+received);
      callback(received);
    });
  }



  /// Initializes the connection. Socket starts listening to server for data
  /// 'callback' function will be called when 'eom' is received
  ///  * @param  the timeOut  amount of time to attempt the connection in milliseconds
  ///  * @param  the eom  sequence of characters at the end of each single message
  ///  * @param  the callback  function called when received a message. It must take a 'String' as param which is the message received
  ///  * @param  the attempts  number of attempts before stop trying to connect. Default is 1
  connectEOM(int timeOut, String eom,  Function callback, {int attempts=1}) async{
    if(_ipAddress==null){
      print("Class not initialized. You must call the constructor!");
      return;
    }
    int k=1;
    while(k<=attempts){
      try{
        _server = await Socket.connect(_ipAddress, _portAddress, timeout: new Duration(milliseconds: timeOut));
        break;
      }catch(Exception){
        _printData(k.toString()+" attempt: Socket not connected (Timeout reached)");
        if(k==attempts){
          return;
        }
      }
      k++;
    }
    _connected=true;
    _printData("Socket successfully connected");
    StringBuffer message=new StringBuffer();
    _server.listen((List<int> event) async {
      String received=(utf8.decode(event));
      message.write(received);
      if(received.contains(eom)){
        _printData("Message received: "+message.toString());

        List<String> messages=message.toString().split(eom);
        if(!received.endsWith(eom)){
          message.clear();
          message.write(messages.last);
          messages.removeLast();
        }
        else{
          message.clear();
        }
        for(String m in messages){
          callback(m);
        }
      }
    });
  }

  /// Stops the connection and close the socket
  void disconnect(){
    if(_server!=null){
      try{
        _server.close();
        _printData("Socket disconnected successfully");
      }catch(exception){
        print("ERROR"+exception);
      }
    }
    _connected=false;
  }

  /// Checks if the socket is connected
  bool isConnected(){
    return _connected;
  }

  /// Sends a message to server. Make sure to have established a connection before calling this method
  /// Message will be sent as 'message'
  ///  * @param  message  message to send to server
  void sendMessage(String message) async{
    if(_server!=null &&_connected){
      _server.add(utf8.encode(message));
      _printData("Message sent: "+message);
    }else{
      print("Socket not initialized before sending message! Make sure you have already called the method 'connect()'");
    }
  }

  /// Sends a message to server. Make sure to have established a connection before calling this method
  /// Message will be sent as 'message'+'eom'
  ///  * @param  message  the message to send to server
  ///  * @param  eom  the end of message to send to server
  void sendMessageEOM(String message,String eom) async{
    if(_server!=null &&_connected){
      _server.add(utf8.encode(message+eom));
      _printData("Message sent: "+message+eom);
    }else{
      print("Socket not initialized before sending message! Make sure you have already called the method 'connect()'");
    }
  }
/*
  /// Send message to server with a command. Make sure to have established a connection before calling this method. Never use this if the connection is established with the 'simpleConnnect()' method
  /// Message will be sent as 'command'+'separator'+'message'+'separator'+'eom'
  ///  * @param  message  message to send to server
  ///  * @param  command  tells the server what to do with the message
  void sendMessageWithCommand(String message, String command) async{
    if(_server!=null){
      _server.add(utf8.encode(command+_separator+message+_separator+_eom));
      _printData("Message sent: "+command+_separator+message+_separator+_eom);
    }else{
      print("Socket not initialized before sending message! Make sure you have alreadt called the method 'connect()'");
    }
  }*/


  /// Test the connection. It will try to connect to the endpoint and if it does, it will disconnect and return 'true' (otherwise false)
  ///  * @param  timeOut  the amount of time to attempt the connection in milliseconds
  ///  * @param  attempts  the number of attempts before stop trying to connect. Default is 1.
  Future<bool> canConnect(int timeOut,{int attempts=1}) async{
    if(_ipAddress==null){
      print("Class not initialized. You must call the constructor!");
      this.disconnect();
      return false;
    }
    int k=1;
    while(k<=attempts){
      try{
        _server = await Socket.connect(_ipAddress, _portAddress, timeout: new Duration(milliseconds: timeOut));
        this.disconnect();
        return true;
      }catch(Exception){
        _printData(k.toString()+" attempt: Socket not connected (Timeout reached)");
        if(k==attempts){
          this.disconnect();
          return false;
        }
      }
      k++;
    }
    this.disconnect();
    return false;
  }

  void _printData(String data){
    if(_logPrintEnabled){
      print(data);
    }
  }
}

example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:tcp_socket_connection/tcp_socket_connection.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
        visualDensity: VisualDensity.adaptivePlatformDensity,
      ),
      home: MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);
  final String title;
  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  //Instantiating the class with the Ip and the PortNumber
  TcpSocketConnection socketConnection=TcpSocketConnection("192.168.1.113", 10251);

  String message="";

  @override
  void initState() {
    super.initState();
    startConnection();
  }

  //receiving and sending back a custom message
  void messageReceived(String msg){
    setState(() {
      message=msg;
    });
    socketConnection.sendMessage("MessageIsReceived :D ");
  }

  //starting the connection and listening to the socket asynchronously
  void startConnection() async{
    socketConnection.enableConsolePrint(true);    //use this to see in the console what's happening
    if(await socketConnection.canConnect(5000, attempts: 3)){   //check if it's possible to connect to the endpoint
      await socketConnection.connect(5000, messageReceived, attempts: 3);
    }
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Center(
        child: Text(
          'You have received '+ message,
        ),
      ),
    );
  }
}
